upup/pkg/fi/cloudup/gce/tpm/gcetpmverifier/tpmverifier.go

```diff
@@ -0,0 +1,241 @@
+	}
+
+	return &tpmVerifier{
+		opt:           *opt,
+		computeClient: computeClient,
+	}, nil
+}
+
+var _ bootstrap.Verifier = &tpmVerifier{}
+
+func (v *tpmVerifier) Verify(ctx context.Context, token string, body []byte) (*bootstrap.VerifyResult, error) {
+	bootstrapRequest := &nodeup.BootstrapRequest{}
+	if err := json.Unmarshal(body, bootstrapRequest); err != nil {
+		return nil, fmt.Errorf("failed to decode request: %w", err)
+	}
+
+	claim := bootstrapRequest.Claim
+	if claim == nil {
+		return nil, fmt.Errorf("claim is required")
+	}
+
+	node := claim.Node
+	if claim.Node == nil {
+		return nil, fmt.Errorf("node is required")
+	}
+
+	if node.GCPProjectID == "" {
+		return nil, fmt.Errorf("gcpProjectID is required")
+	}
+	if node.Zone == "" {
+		return nil, fmt.Errorf("zone is required")
+	}
+	if node.Instance == "" {
+		return nil, fmt.Errorf("instance is required")
+	}
+
+	if node.GCPProjectID != v.opt.ProjectID {
+		return nil, fmt.Errorf("projectID does not match expected: got %q, want %q", node.GCPProjectID, v.opt.ProjectID)
+	}
+
+	instance, err := v.computeClient.Instances.Get(node.GCPProjectID, node.Zone, node.Instance).Do()
+	if err != nil {
+		if isNotFound(err) {
+			return nil, fmt.Errorf("unable to find instance in compute API: %w", err)
+		}
+		return nil, fmt.Errorf("error fetching instance from compute API: %w", err)
+	}
+
+	if !strings.HasPrefix(lastComponent(instance.Zone), v.opt.Region+"-") {
+		return nil, fmt.Errorf("instance was in zone %q, expected region %q", instance.Zone, v.opt.Region)
+	}
+
+	// Verify node is in our cluster
+	clusterName := ""
+	instanceGroupName := ""
+	for _, item := range instance.Metadata.Items {
+		switch item.Key {
+		case gce.MetadataKeyInstanceGroupName:
+			instanceGroupName = fi.StringValue(item.Value)
+		case "cluster-name":
+			clusterName = fi.StringValue(item.Value)
+		}
+	}
+
+	if clusterName == "" {
+		return nil, fmt.Errorf("could not determine cluster for instance %s", instance.SelfLink)
+	}
+
+	if clusterName != v.opt.ClusterName {
+		return nil, fmt.Errorf("clusterName does not match expected: got %q, want %q", clusterName, v.opt.ClusterName)
+	}
+	if instanceGroupName == "" {
+		return nil, fmt.Errorf("could not determine instance group for instance %s", instance.SelfLink)
+	}
+
+	// TODO can we include the attestation certificate (signed by GCE) in the claim, rather than getting it from the API?
+	tpmSigningKey, err := v.getTPMSigningKey(ctx, bootstrapRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// Verify node
+	{
+		nodeClaim, err := json.Marshal(node)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal node claim: %w", err)
+		}
+		if !verifySignature(tpmSigningKey, nodeClaim, claim.NodeSignature) {
+			return nil, fmt.Errorf("failed to verify claim signature for node: %w", err)
+		}
+	}
+
+	sans, err := GetInstanceCertificateAlternateNames(instance)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
This completely fails to verify the request body!!! An attacker obtaining data signed by the TPM that deserializes to a nodeClaim that identifies the TPM's instance could use it to obtain certs for keys of the attacker's choice.

Reply:
Well spotted, and fixed now!